File: NRDB/DeckManager.m
//
//  DeckManager.m
//  NRDB
//
//  Created by Gereon Steffens on 18.03.13.
//  Copyright (c) 2014 Gereon Steffens. All rights reserved.
//

#import "DeckManager.h"
#import "Deck.h"
#import "SettingsKeys.h"

#define SAVE_DECK  @"deck_%d"

@implementation DeckManager

// save deck
+(void) saveDeck:(Deck*)deck
{
    if (deck.filename == nil)
    {
        deck.filename = [DeckManager pathForRole:deck.role];
    }
        
    // NSLog(@"saving deck %@ (%@)", deck.name, deck.identity.name);
    
    // Set up the encoder and storage for the saved state data
    NSMutableData* savedData = [NSMutableData data];
    NSKeyedArchiver* encoder = [[NSKeyedArchiver alloc] initForWritingWithMutableData:savedData];
    
    [encoder encodeObject:deck forKey:@"deck"];
    
    // Finish encoding and write to the file
    [encoder finishEncoding];
    
    BOOL saveOk = [savedData writeToFile:deck.filename atomically:YES];
    if (!saveOk)
    {
        [DeckManager removeFile:deck.filename];
        return;
    }
    
    if (deck.dateCreated)
    {
        NSDictionary *attrs = @{ NSFileCreationDate: deck.dateCreated };
        NSError *error;
        NSFileManager* fileMgr = [NSFileManager defaultManager];
        
        [fileMgr setAttributes:attrs ofItemAtPath:deck.filename error: &error];
    }
}

// load a Deck
+(Deck*) loadDeckFromPath:(NSString *)filename
{
    // Check to see if the saved state file exists and if so, load it
    NSData* savedData = [NSData dataWithContentsOfFile:filename];
    if (savedData == nil)
    {
        // no saved data
        return nil;
    }
    
    // decode the saved state
    NSKeyedUnarchiver* decoder = [[NSKeyedUnarchiver alloc] initForReadingWithData:savedData];
    Deck* deck = [decoder decodeObjectForKey:@"deck"];
    deck.filename = filename;
    
    // get the last modification date
    NSFileManager* fm = [NSFileManager defaultManager];
    NSDictionary* attrs = [fm attributesOfItemAtPath:filename error:nil];
    
    if (attrs != nil)
    {
        NSDate *date = (NSDate*)[attrs objectForKey:NSFileModificationDate];
        deck.lastModified = date;
        
        date = (NSDate*)[attrs objectForKey:NSFileCreationDate];
        deck.dateCreated = date;
    }
    
    return deck;
}

+(NSString*) directoryForRole:(NRRole)role
{
    NSAssert(role != NRRoleNone, @"wrong role");
    
    NSString* roleDirectory = role == NRRoleRunner ? @"runnerDecks" : @"corpDecks";
    
    NSArray* paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString* documentsDirectory = [paths objectAtIndex:0];
    NSString* directory = [documentsDirectory stringByAppendingPathComponent:roleDirectory];
    
    if (![[NSFileManager defaultManager] fileExistsAtPath:directory])
    {
        [[NSFileManager defaultManager] createDirectoryAtPath:directory withIntermediateDirectories:NO attributes:nil error:nil];
    }
    
    return directory;
}

// get a new filename
+(NSString*) pathForRole:(NRRole)role
{
    NSString* directory = [DeckManager directoryForRole:role];
    NSString* filename = [NSString stringWithFormat:@"deck-%d.anr", [DeckManager nextFileId]];
    NSString* path = [directory stringByAppendingPathComponent:filename];
    
    // NSLog(@"path: %@", path);
    return path;
}

+(NSMutableArray*) decksForRole:(NRRole)role
{
    if (role == NRFactionNone)
    {
        NSMutableArray* decks = [self readDecksForRole:NRRoleRunner];
        [decks addObjectsFromArray:[self readDecksForRole:NRRoleCorp]];
        return decks;
    }
    else
    {
        return [self readDecksForRole:role];
    }
}

+(NSMutableArray*) readDecksForRole:(NRRole)role
{
    NSString* directory = [DeckManager directoryForRole:role];
    NSArray *dirContents = [[NSFileManager defaultManager] contentsOfDirectoryAtPath:directory error:nil];
    
    NSMutableArray* decks = [NSMutableArray array];
    
    for (NSString* file in dirContents)
    {
        NSString* pathname = [directory stringByAppendingPathComponent:file];
        
        @try
        {
            Deck* deck = [DeckManager loadDeckFromPath:pathname];
            
            [decks addObject:deck];
        }
        @catch (NSException* ex)
        {
            NSLog(@"caught exception loading %@", pathname);
            NSLog(@"exception was: %@ %@", ex.name, ex.reason);
            
            #if RELEASE
            [DeckManager removeFile:pathname];
            #endif
        }
    }
    
    return decks;
}

+(void) removeAll
{
    NSArray* roles = @[ @(NRRoleRunner), @(NRRoleCorp) ];
    
    NSFileManager* fileMgr = [NSFileManager defaultManager];
    NSError* error = nil;
    
    for (NSNumber* role in roles)
    {
        NSString* directory = [DeckManager directoryForRole:[role intValue]];
        NSArray *dirContents = [fileMgr contentsOfDirectoryAtPath:directory error:nil];

        for (NSString* file in dirContents)
        {
            NSString* pathname = [directory stringByAppendingPathComponent:file];
            
            // NSLog(@"removing %@", pathname);
            [fileMgr removeItemAtPath:pathname error:&error];
        }
    }
}

+(int) nextFileId
{
    NSUserDefaults* defaults = [NSUserDefaults standardUserDefaults];
    
    NSNumber* num = [defaults objectForKey:FILE_SEQ];
    int fileId = [num intValue];
    [defaults setObject:@(fileId+1) forKey:FILE_SEQ];
    [defaults synchronize];
    
    return fileId;
}


// remove save state file
+(void) removeFile:(NSString*)pathName
{
    // NSLog(@"removing %@", pathName);
    
    NSError* error;
    NSFileManager *fileMgr = [NSFileManager defaultManager];
    [fileMgr removeItemAtPath:pathName error:&error];
}

// reset a deck's last modification timestamp
+(void) resetModificationDate:(Deck*)deck
{
    if (deck.filename && deck.lastModified)
    {
        NSDictionary *attrs = @{ NSFileModificationDate: deck.lastModified };
        NSError *error;
        NSFileManager* fileMgr = [NSFileManager defaultManager];

        [fileMgr setAttributes:attrs ofItemAtPath:deck.filename error: &error];
    }
}

@end

File: NetDeck/DeckManager.swift
//
//  DeckManager.swift
//  NetDeck
//
//  Created by Gereon Steffens on 05.12.15.
//  Copyright © 2018 Gereon Steffens. All rights reserved.
//

import Foundation
import SwiftyUserDefaults

class DeckManager {
    
    static let cache: NSCache<NSString, Deck> = {
        let c = NSCache<NSString, Deck>()
        c.name = "deckCache"
        return c
    }()
    
    static func saveDeck(_ deck: Deck, keepLastModified: Bool) {
        if deck.filename == nil {
            deck.filename = DeckManager.pathForRole(deck.role)
        }
        
        let filename = deck.filename!
        // print("save \(deck.name) to disk")
        let data = NSMutableData()
        let encoder = NSKeyedArchiver(forWritingWith: data)
        
        encoder.encode(deck, forKey: "deck")
        encoder.finishEncoding()
        
        let saveOk = data.write(toFile: filename, atomically: true)
        if !saveOk {
            DeckManager.removeFile(filename)
            return
        }
        
        let fileMgr = FileManager.default
        if let date = deck.dateCreated {
            let attrs = [ FileAttributeKey.creationDate: date ]
            
            _ = try? fileMgr.setAttributes(attrs, ofItemAtPath: filename)
        }
        
        if keepLastModified && deck.lastModified != nil {
            let attrs = [ FileAttributeKey.modificationDate: deck.lastModified! ]
            _ = try? FileManager.default.setAttributes(attrs, ofItemAtPath: deck.filename!)
        } else {
            // update the in-memory lastModified date
            deck.lastModified = Date()
        }

        // store the deck in our cache
        DeckManager.cache.setObject(deck, forKey: filename as NSString)
    }
    
    static func removeFile(_ pathname: String) {
        DeckManager.cache.removeObject(forKey: pathname as NSString)
        _ = try? FileManager.default.removeItem(atPath: pathname)
    }
    
    static func decksForRole(_ role: Role) -> [Deck] {
        if role == .none {
            var decks = [Deck]()
            decks.append(contentsOf: loadDecksForRole(.runner))
            decks.append(contentsOf: loadDecksForRole(.corp))
            return decks
        }
        return loadDecksForRole(role)
    }
    
    private static func loadDecksForRole(_ role: Role) -> [Deck] {
        var decks = [Deck]()
        
        let dir = directoryForRole(role)
        if let dirContents = try? FileManager.default.contentsOfDirectory(atPath: dir) {
            for file in dirContents {
                let path = dir.appendPathComponent(file)
                if let deck = loadDeckFromPath(path) {
                    decks.append(deck)
                }
            }
        }
        return decks
    }
    
    static func loadDeckFromPath(_ path: String, useCache: Bool = true) -> Deck? {
        if useCache {
            if let cachedDeck = DeckManager.cache.object(forKey: path as NSString) {
                return cachedDeck
            }
        }
        
        if let data = try? Data(contentsOf: URL(fileURLWithPath: path)) {
            let decoder = NSKeyedUnarchiver(forReadingWith: data)
            if let deck = decoder.decodeObject(forKey: "deck") as? Deck {
                deck.filename = path

                if let attrs = try? FileManager.default.attributesOfItem(atPath: path) {
                    deck.lastModified = attrs[FileAttributeKey.modificationDate] as? Date
                    deck.dateCreated = attrs[FileAttributeKey.creationDate] as? Date
                }

                DeckManager.cache.setObject(deck, forKey: path as NSString)
                return deck
            }
        }
        return nil
    }
    
    static func numberOfDecks() -> Int {
        let roles = [Role.runner, .corp]
        
        let decks = roles.reduce(0) {
            let dir = directoryForRole($1)
            let files = try? FileManager.default.contentsOfDirectory(atPath: dir)
            return $0 + (files?.count ?? 0)
        }
        
        return decks
    }
    
    static func directoryForRole(_ role: Role) -> String {
        assert(role != .none, "wrong role")
        
        let roleDir = role == .runner ? "runnerDecks" : "corpDecks"
        let paths = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)
        let dir = paths[0].appendPathComponent(roleDir)
        
        let fileMgr = FileManager.default
        if !fileMgr.fileExists(atPath: dir) {
            _ = try? fileMgr.createDirectory(atPath: dir, withIntermediateDirectories: true, attributes: nil)
        }
        return dir
    }
    
    static func pathForRole(_ role: Role) -> String {
        let dir = directoryForRole(role)
        let file = String(format:"deck-%d.anr", nextFileSequence())
        let path = dir.appendPathComponent(file)
        
        return path
    }
    
    static func fileSequence() -> Int {
        return Defaults[.fileSequence]
    }
    
    private static func nextFileSequence() -> Int {
        let seq = Defaults[.fileSequence]
        Defaults[.fileSequence] += 1
        return seq
    }
    
    static func flushCache() {
        DeckManager.cache.removeAllObjects()
    }
}
